Accept a login hint in the Caddyfile's authorize block

PedroDBFlores opened the pull request:
For some OAuth2 providers, there is a possibility of providing a `login_hint` parameter that will prefill the user's login on that provider (usually email).

For some routes I would like to be able to send the user to a login page with it's email prefilled, for this I thought of adding a `set login hint my_hint` parameter to the `authorize` block on the Caddyfile. The idea is that this login hint is then forwarded to the authorizer, and handled there.

You can set the login hint by directly setting the value there, but it also supports getting a query parameter from the requested uri, for example if CaddyFile has `set login hint {http.request.uri.query.email}`  then going to http://acme.org/login?email=foo@bar.com will take foo@bar.com as a login hint.

I ran all the tests and there already was a failing test from the start on `TestParseCryptoKeyConfigs`, besides that all is green.

Commit:
I hereby consent to the Individual CLA provided in assets/cla/individual_cla.md

Accept a login hint in the authorize{} block

Forward the login hint to the authorizer

greenpau commented on pkg/handlers/redirect.go:
```diff
@@ -48,18 +54,23 @@ func HandleHeaderRedirect(w http.ResponseWriter, r *http.Request, opts map[strin
 	authURLPath = opts["auth_url_path"].(string)
 	authRedirectQueryDisabled := opts["auth_redirect_query_disabled"].(bool)
 	redirectParameter = opts["redirect_param"].(string)
+	loginHint = ""
```
@PedroDBFlores , the `loginHint` is not being used.

greenpau commented on pkg/handlers/redirect.go:
```diff
@@ -48,18 +54,23 @@ func HandleHeaderRedirect(w http.ResponseWriter, r *http.Request, opts map[strin
 }
 
-func redirectParameters(w http.ResponseWriter, r *http.Request, opts map[string]interface{}) (authURLPath, sep, redirectParameter, redirectURL string, redirect bool) {
+func redirectParameters(_ http.ResponseWriter, r *http.Request, opts map[string]interface{}) (authURLPath, sep, redirectParameter, redirectURL string, loginHint string, redirect bool) {
```
@PedroDBFlores , no need to change the signature. The `login_hint`, if any would be in `opts`.

greenpau commented on pkg/handlers/redirect.go:
```diff
@@ -22,17 +22,23 @@ import (
+// HandleHeaderRedirect redirect the requests to configured auth URL by setting Location header and sending 302.
 func HandleHeaderRedirect(w http.ResponseWriter, r *http.Request, opts map[string]interface{}) {
-	authURLPath, sep, redirectParameter, redirectURL, redirect := redirectParameters(w, r, opts)
+	authURLPath, sep, redirectParameter, redirectURL, loginHint, redirect := redirectParameters(w, r, opts)
```
@PedroDBFlores , the the hint is already in `opts`. Thus, no need to create the `loginHint` variable.

Commit:
Merge branch 'main' of https://github.com/greenpau/caddy-authorize into greenpau-main

Commit:
Merge branch 'greenpau-main'

Commit:
I hereby consent to the Individual CLA provided in assets/cla/individual_cla.md

Change to enable login_hint instead of providing it

Comes through query param `login_hint` if unauthenticated

If previously authenticated, Email is used for login_hint

PedroDBFlores commented:
@greenpau I've made the changes you suggested with using the `enable login hint` method. Check my latest commit and let me know what you think.

greenpau commented on pkg/utils/validate/login_hint.go:
```diff
@@ -0,0 +1,10 @@
+	"net/mail"
+)
+
+func ValidateLoginHint(loginHint string) bool {
```
@PedroDBFlores , should return error, not bool.

greenpau commented on pkg/utils/validate/login_hint_test.go:
```diff
@@ -0,0 +1,35 @@
+
+import "testing"
+
+func TestLoginHint(t *testing.T) {
```
@PedroDBFlores , please see the following testing pattern and mimic it. https://github.com/greenpau/caddy-authorize/blob/main/pkg/shared/idp/config_test.go

greenpau commented on pkg/authz/authorizer.go:
```diff
@@ -198,9 +200,21 @@ func (m Authorizer) Authenticate(w http.ResponseWriter, r *http.Request, upstrea
+				queryHasLoginHint := r.URL.Query().Has("login_hint")
+				var loginHint = ""
+
+				if queryHasLoginHint {
```
@PedroDBFlores , let's implement `switch` here. Also, let's not use `queryHasLoginHint` and `loginHint:

```
switch {
case r.URL.Query().Has("login_hint"):
  redirOpts["login_hint"] = r.URL.Query().Get("login_hint")
case !r.URL.Query().Has("login_hint") && usr != nil:
  # What if usr.Claims.Email (should be using usr.Authenticator.LoginHint instead; putting commit as I type) is empty?   Perhaps checking it might work better.
  redirOpts["login_hint"] =  usr.Authenticator.LoginHint  
}
```

greenpau commented on pkg/authz/authorizer.go:
```diff
@@ -198,9 +200,21 @@ func (m Authorizer) Authenticate(w http.ResponseWriter, r *http.Request, upstrea
+					loginHint = usr.Claims.Email
+				}
+
+				if err := m.handleLoginHint(loginHint, redirOpts); err != nil {
```
@PedroDBFlores , with the `loginHint` is in `redirOpts`, you only need to pass `redirOpts`. Also, do not throw an error. Rather use `p.logger.Warn()` to indicate that the hint had an issue.

greenpau commented on pkg/authz/authorizer.go:
```diff
@@ -263,39 +277,14 @@ func (m Authorizer) Authenticate(w http.ResponseWriter, r *http.Request, upstrea
 	return userIdentity, true, nil
 }
 
-func (m Authorizer) parseLoginHint(r *http.Request, redirOpts map[string]interface{}) {
```
@PedroDBFlores , the `Authorizer` should be a pointer.

```
 func (m Authorizer) parseLoginHint(r *http.Request, redirOpts map[string]interface{}) {
 func (m *Authorizer) parseLoginHint(r *http.Request, redirOpts map[string]interface{}) {
```

greenpau commented on pkg/authz/authorizer.go:
```diff
@@ -263,39 +277,14 @@ func (m Authorizer) Authenticate(w http.ResponseWriter, r *http.Request, upstrea
-		redirOpts["login_hint"] = loginHint
-	} else {
-		redirOpts["login_hint"] = m.LoginHint
+func (m Authorizer) handleLoginHint(loginHint string, redirOpts map[string]interface{}) error {
```
@PedroDBFlores , the `Authorizer` should be a pointer.

```
 func (m Authorizer) parseLoginHint(r *http.Request, redirOpts map[string]interface{}) {
 func (m *Authorizer) parseLoginHint(r *http.Request, redirOpts map[string]interface{}) {
```

Commit:
I hereby consent to the Individual CLA provided in assets/cla/individual_cla.md

Support login hint validators

greenpau commented on caddyfile.go:
```diff
@@ -329,13 +330,35 @@ func parseCaddyfile(h httpcaddyfile.Helper) (*authz.Authorizer, error) {
+						remainingArguments = strings.TrimPrefix(remainingArguments, "with ")
+						validationArguments := strings.Split(remainingArguments, " ")
+
+						fmt.Sprintf("%s", validationArguments)
```
@PedroDBFlores , remove

greenpau commented on caddyfile.go:
```diff
@@ -329,13 +330,35 @@ func parseCaddyfile(h httpcaddyfile.Helper) (*authz.Authorizer, error) {
 				args := strings.Join(h.RemainingArgs(), " ")
-				switch args {
-				case "js redirect":
+				fmt.Sprintf("%s", args)
```
@PedroDBFlores , remove.

greenpau commented on caddyfile.go:
```diff
@@ -329,13 +330,35 @@ func parseCaddyfile(h httpcaddyfile.Helper) (*authz.Authorizer, error) {
 					p.LoginHintEnabled = true
+					remainingArguments := strings.TrimPrefix(args, "login hint ")
+
+					if strings.HasPrefix(remainingArguments, "with") {
```
Please use `switch`.

Commit:
I hereby consent to the Individual CLA provided in assets/cla/individual_cla.md

Use switch instead

Remove dangling sprintf

greenpau commented on caddyfile.go:
```diff
@@ -356,9 +352,12 @@ func parseCaddyfile(h httpcaddyfile.Helper) (*authz.Authorizer, error) {
+						break
+					default:
 						p.LoginHintValidators = []string{"alphanumeric"}
+						break
```
Let’s the default be all three validations. Any of them passes, it is good to go.

Commit:
I hereby consent to the Individual CLA provided in assets/cla/individual_cla.md

Create "disabled" login hint validator

Commit:
I hereby consent to the Individual CLA provided in assets/cla/individual_cla.md

Add inheritance of login hint validators from primary instance

greenpau commented on pkg/handlers/js_redirect.go:
```diff
@@ -44,7 +44,7 @@ var jsRedirTmpl = template.Must(template.New("js_redir").Parse(`
 // with javascript doing the real redirection.
 func HandleJSRedirect(w http.ResponseWriter, r *http.Request, opts map[string]interface{}) {
-	authURLPath, sep, redirectParameter, redirectURL, redirect := redirectParameters(w, r, opts)
+	authURLPath, sep, redirectParameter, redirectURL, _, redirect := redirectParameters(w, r, opts)
```
@PedroDBFlores , @lennartschoch , please do not change the number of returned values from `redirectParameters()`.

Please use the `opt` map to pass `loginHint` around.

greenpau commented on pkg/handlers/redirect.go:
```diff
@@ -48,18 +54,23 @@ func HandleHeaderRedirect(w http.ResponseWriter, r *http.Request, opts map[strin
 }
 
-func redirectParameters(w http.ResponseWriter, r *http.Request, opts map[string]interface{}) (authURLPath, sep, redirectParameter, redirectURL string, redirect bool) {
+func redirectParameters(_ http.ResponseWriter, r *http.Request, opts map[string]interface{}) (authURLPath, sep, redirectParameter, redirectURL string, loginHint string, redirect bool) {
```
@PedroDBFlores , @lennartschoch , please do the calculation of the `finalURL` here, in `redirectParameters()` function (not in `HandleHeaderRedirect()`.

PedroDBFlores replied:
I've done the requested changes on the last two commits 👍

Commit:
I hereby consent to the Individual CLA provided in assets/cla/individual.md

Remove loginHint from redirectParameters return

Commit:
Add login_hint query parameter in redirectParameters function

Commit:
Add comment to LoginHint function

PedroDBFlores commented:
@greenpau Any other remarks on this pull request?

greenpau commented:
@PedroDBFlores , @lennartschoch , this PR would have to move. Let’s resume after I finish https://github.com/greenpau/caddy-authorize/issues/98

greenpau commented:
@PedroDBFlores , @lennartschoch , I am all done migrating the code! 👍 

All Caddyfile related things would require opening a PR at https://github.com/greenpau/caddy-security

Anything else, you would need to open it here: https://github.com/greenpau/aaasf . Specifically, https://github.com/greenpau/aaasf/tree/main/pkg/authz
